src/ActionParameterHandler.ts

```diff
@@ -103,43 +105,83 @@ export class ActionParameterHandler<T extends BaseDriver> {
     /**
      * Normalizes parameter value.
      */
-    protected normalizeParamValue(value: any, param: ParamMetadata): Promise<any>|any {
+    protected async normalizeParamValue(value: any, param: ParamMetadata): Promise<any> {
         if (value === null || value === undefined)
             return value;
 
-        switch (param.targetName) {
+        // if param value is an object and param type match, normalize its string properties
+        if (typeof value === "object" && ["queries", "headers", "path-params", "cookies"].some(paramType => paramType === param.type)) {
+            Object.keys(value).map(key => {
+                const keyValue = (value as any)[key];
+                if (typeof keyValue === "string") {
+                    const ParamType = Reflect.getMetadata("design:type", param.targetType.prototype, key);
+                    if (ParamType) {
+                        const typeString = ParamType.name.toLowerCase(); // reflected type is always constructor-like (?)
+                        (value as any)[key] = this.normalizeStringValue(keyValue, param.name, typeString);
+                    }
+                }
+            });
+        }
+        // if value is a string, normalize it to demanded type
+        else if (typeof value === "string") {
+            switch (param.targetName) {
+                case "number":
+                case "string":
+                case "boolean":
+                case "date":
+                    return this.normalizeStringValue(value, param.name, param.targetName);
+            }
+        }
+
+        // if target type is not primitive, transform and validate it
+        if ((["number", "string", "boolean"].indexOf(param.targetName) === -1)
+            && (param.parse || param.isTargetObject)
+        ) {
+            value = this.parseValue(value, param);
+            value = this.transformValue(value, param);
+            value = this.validateValue(value, param); // note this one can return promise
+        }
+
+        return value;
+    }
+
+    /**
+     * Normalizes string value to number or boolean.
+     */
+    protected normalizeStringValue(value: string, parameterName: string, parameterType: string) {
+        switch (parameterType) {
             case "number":
-                if (value === "") return undefined;
-                return +value;
+                if (value === "") {
+                    throw new ParamNormalizationError(value, parameterName, parameterType);
```

Comment:
in `ParamNormalizationError` I don't like `Normalization` word. It does not tell user so much, user may think - what is "normalization?". Normalization is more implementation-specific, what is going on in this code. I think it will be better user experience if we replace Normalization with at least Validation

Reply:
Normalization is _"process that makes something more normal or regular"_. However the better name would be transforming, because we transform string to numbers or boolean, however I don't know what error "class-transformer" throws (if any?).

Reply:
> Normalization is a "process that makes something more normal or regular". 

right, thats why I called this method "normalize". But its routing-controllers internals, we need something more user-friendly for users since they will see this message. Same with tranformation. We shall tell users "dude you have a wrong parameter" instead of "dude transormation of your parameter failed" or "dude normalization of your parameter failed".

Reply:
It's just an internal name. Have you checked the source code of this error?
```
`Given parameter ${parameterName} is invalid. Value (${JSON.stringify(value)}) cannot be parsed into ${parameterType}.`
```
Isn't it a `user-friendly` message?

Reply:
> just internal name

no, error name is used as well.

> Isn't it a user-friendly message?

yes, it is. And we can clearly admin that class name is a short name of message inside it. And message inside it does not tell anything about normalization or transformation

Reply:
So you suggest that it should be "InvalidParamError"?